client/models.py

```diff
@@ -119,8 +161,66 @@ def get_active_certificates(self):
 
 
 class VolunteerCertificate(models.Model):
+    CERTIFICATE_TEXT_COLOR = (3, 8, 12)  # (R, G, B)
+    CERTIFICATE_TEXT_SIZE = 40
+    CERTIFICATE_TEXT_POSITION = (700, 200)
+    IMAGE_PATH = 'certificates/{id}.png'
+    DOWNLOAD_LINK_EXPIRATION_SECONDS = 600
+
+    def _certificate_image_path(self, filename):
+        return type(self).IMAGE_PATH.format(id=self.id)
+
     volunteer = models.ForeignKey(Volunteer, on_delete=models.CASCADE, related_name='certificates', null=False)
     expiration_date = models.DateField(default=date.today)
+    image = models.FileField(blank=True, upload_to=_certificate_image_path)
+
+    def create_image(self):
+        volunteer = self.volunteer
+        tag_filename = finders.find('client/tag.jpeg')
+        font_filename = finders.find('client/fonts/BN Amnesia.ttf')
+        photo = None
+        try:
+            photo = Image.open(tag_filename)
+            drawing = ImageDraw.Draw(photo)
+            font = ImageFont.truetype(font_filename, size=type(self).CERTIFICATE_TEXT_SIZE)
+
+            lines_to_insert = [
+                f'שם מתנדב: {volunteer.first_name} {volunteer.last_name}',
+                f'תעודת זהות: {volunteer.tz_number}',
+                f'תוקף התעודה: {self.expiration_date}',
+                f'מספר תעודה: {self.id}',
+            ]
+
+            drawing.text(
+                type(self).CERTIFICATE_TEXT_POSITION,
+                get_display('\n'.join(lines_to_insert)),
+                fill=type(self).CERTIFICATE_TEXT_COLOR,
+                font=font,
+                align='right'
+            )
+            with io.BytesIO() as output:
+                photo.save(output, format='png')
+                self.image.save(type(self).IMAGE_PATH.format(id=self.id), ContentFile(output.getvalue()))
+        finally:
+            if photo is not None:
+                photo.close()
+
+    def update_image_if_nonexistent(self):
+        if not self.image.name:
+            self.create_image()
+
+    @property
+    def image_download_url(self):
+        s3 = boto3.client('s3')
+        return s3.generate_presigned_url(
+            'get_object',
+            Params={
+                'Bucket': 'levechad-media-bucket',
+                'Key': 'media/{}'.format(type(self).IMAGE_PATH.format(id=self.id)),
+                'ResponseContentDisposition': 'attachment;filename={}'.format(f'{self.id}.png'),
+            },
+            ExpiresIn=type(self).DOWNLOAD_LINK_EXPIRATION_SECONDS
+        )
```

Comment:
why is it needed?

Reply:
you can always use certificate.image.url

Reply:
But that isn't a download link.
We used to use the download attribute on <a> tags, but this only works in Firefox and Chrome if the image is not from an external URL, which it now is (served from s3, not from django).
S3 has presigned URLs that expire after a while where you can give access to a resource using a URL, and you can also set the content disposition, which makes the users download the file.